### PR TITLE
Move search implementation from generic `IndexView` to `BaseListingView`

### PR DESCRIPTION
**Pull request by laymonage**
Quite nuanced, but I think the cleanups here are worth it. See each commit message for more details.

### CODE REVIEW

**Comment by laymonage** [comment]
Just rebased.

File: wagtail/admin/views/generic/base.py [hunk 1]
```diff
@@ -191,8 +196,13 @@ class BaseListingView(WagtailAdminTemplateMixin, BaseListView):
     index_url_name = None
     index_results_url_name = None
     page_kwarg = "p"
+    is_searchable = None  # Subclasses must explicitly set this to True to enable search
```
[the remark this answers is not in the document]
Comment (gasman) [reply wagtail/admin/views/generic/base.py:199]: That sounds reasonable. As far as I can see the only reason BaseListingView needs to care about the existence of a `model` property at all is that Django's `BaseListView.get_queryset` uses `self.model._default_manager.all()` as a default, and it's highly likely that subclasses will override that, at which point we can't assume anything about what it means...

File: wagtail/admin/templates/wagtailadmin/generic/listing_results.html [hunk 2]
```diff
@@ -33,6 +49,23 @@
     {% endblock %}
 {% else %}
     <div class="nice-padding w-mt-8">
-        {% block no_results_message %}<p>{% trans "There are no results." %}</p>{% endblock %}
+        {% fragment as no_results_message %}
```
[the remark this answers is not in the document]
Comment (laymonage) [reply wagtail/admin/templates/wagtailadmin/generic/listing_results.html]: Sure! I really wanted to do it but wasn't sure if we should make the `index_results.html` diverge, but I think this allows us to display more specific messages for `BaseListingView` subclasses more easily without having to extend the template.

Done in 64dc246.

File: wagtail/admin/views/generic/base.py [hunk 3]
```diff
@@ -212,6 +222,65 @@ def get_breadcrumbs_items(self):
...
+        # Django ORM can still be used on an indexed model by unsetting
+        # search_backend_name and defining search_fields on the view.
+        if class_is_indexed(queryset.model) and self.search_backend_name:
+            search_backend = get_search_backend(self.search_backend_name)
+            if queryset.model.get_autocomplete_search_fields():
+                return search_backend.autocomplete(
+                    self.search_query,
+                    queryset,
+                    fields=self.search_fields,
+                    order_by_relevance=(not self.is_explicitly_ordered),
```
[the remark this answers is not in the document]
Comment (laymonage) [reply wagtail/admin/views/generic/base.py:262]: Note: we technically can "sniff" the model in `get_ordering()` by doing `self.get_queryset().model` without evaluating the queryset, but that feels like a hack.

File: wagtail/admin/templates/wagtailadmin/generic/listing_results.html [hunk 4]
```diff
@@ -49,21 +49,6 @@ <h2 role="alert">
...
-            {% elif is_searching or is_filtering %}
-                {% blocktrans trimmed with model_name=verbose_name_plural asvar no_results_text %}
-                    No {{ model_name }} match your query.
-                {% endblocktrans %}
-            {% else %}
-                {% blocktrans trimmed with model_name=verbose_name_plural asvar no_results_text %}
-                    There are no {{ model_name }} to display.
-                {% endblocktrans %}
-            {% endif %}
-            {{ no_results_text|capfirst }}
```
Comment (gasman) [inline wagtail/admin/templates/wagtailadmin/generic/listing_results.html]: Looks like this `capfirst` has got lost in the move.

Comment (laymonage) [reply wagtail/admin/templates/wagtailadmin/generic/listing_results.html]: Oops, thanks! Rebased and amended as 9d8507b

**Review by gasman — APPROVED** [review]
Thanks @laymonage - happy with this now! I've been able to get a basic listing view with search working by adding the following to base/wagtail_hooks.py in bakerydemo:

```python
class CustomListingView(BaseListingView):
    page_title = "Custom Person Listing"
    search_fields = ["first_name", "last_name"]
    is_searchable = True
    index_url_name = "custom_person"
    index_results_url_name = "custom_person_results"

    def get_base_queryset(self):
        return Person.objects.all()


@hooks.register("register_admin_urls")
def custom_person_view():
    return [
        path(
            "custom_person/",
            CustomListingView.as_view(),
            name="custom_person",
        ),
        path(
            "custom_person/results/",
            CustomListingView.as_view(results_only=True),
            name="custom_person_results",
        )
    ]
```

which I think is a pretty good outcome :-)

Still trying to get my head around the need for fallbacks for listings without breadcrumbs. If there are indeed no such views within Wagtail, and the only way for third-party views to _not_ have breadcrumbs is if they set `_show_breadcrumbs = False` or use the wagtailadmin/generic/listing.html template directly without going though `BaseListingView` - both of which are unsupported - then we should be all clear to eliminate that, right?

**Comment by laymonage** [comment]
Thanks @gasman!

> If there are indeed no such views within Wagtail, and the only way for third-party views to _not_ have breadcrumbs is if they set `_show_breadcrumbs = False` or use the wagtailadmin/generic/listing.html template directly without going though `BaseListingView` - both of which are unsupported - then we should be all clear to eliminate that, right?

Well, technically we _have_ somewhat eliminated the support for `BaseListingView` without breadcrumbs in this PR. Breadcrumbs will always show in any views that use `BaseListingView` and/or `wagtailadmin/generic/listing.html` template – and while forcing `_show_breadcrumbs = False` on the view would disable the breadcrumbs and give the old header back, this is not something specific to `BaseListingView`. It's a result of having the default logic in `wagtailadmin/generic/base.html`:
https://github.com/wagtail/wagtail/blob/954519390753ab47b1f43c05d377496fe38f8c93/wagtail/admin/templates/wagtailadmin/generic/base.html#L8-L20

So there is actually no code left in `BaseListingView`/`wagtailadmin/generic/listing.html` that add explicit support for the non-breadcrumbs version 🙂

My vision is that the view templates should not need to be concerned about overriding the `header` block to add stuff in there (e.g. header buttons, search and filters) in favour of defining them on the view class. Putting the breadcrumbs logic in the `base.html` template allows us to inch closer to that goal as we refactor the views, since we won't need to touch that block for new code, while still providing basic heading support to any views (including non-listing ones) that extend that template and not use the breadcrumbs yet.

Anyway, to recap, what this PR does is:

- Enforce breadcrumbs on all listing views by default.
- Remove the ability to have filters in `BaseListingView`/`wagtailadmin/generic/listing.html` if breadcrumbs are disabled.
  - The reason why we had this before was because the report views were still using it before 6.2. Now, nothing in core relies on this.
  - This ability is now moved to the `wagtailadmin/generic/index.html` template, in case people still use it somehow. Very unlikely, so we in theory can also remove it from the `wagtailadmin/generic/index.html` template. **This is the part that we can clean up** – let me know if we want to! 😉 
- Add the ability to have search in `BaseListingView`/`wagtailadmin/generic/listing.html`.
  - This requires the use of breadcrumbs, because we don't want to support the old header for searching. Hence why we enable breadcrumbs by default, so users don't need to know about it.
  - AFAIK, this view/template never supported searching at all, since we never pass any required variables for searching with the old header. This block is removed in this PR: https://github.com/wagtail/wagtail/blob/954519390753ab47b1f43c05d377496fe38f8c93/wagtail/admin/templates/wagtailadmin/generic/listing.html#L4-L8
  - So any views that used this class/template with searching had to have their own search implementation and/or use a template that has the code for it. This was the case with the page `SearchView`, and we refactor it in this PR.

**Comment by gasman** [comment]
> This ability is now moved to the wagtailadmin/generic/index.html template, in case people still use it somehow. Very unlikely, so we in theory can also remove it from the wagtailadmin/generic/index.html template. This is the part that we can clean up – let me know if we want to! 😉

Given the combination of A) it's very unlikely people have done this, B) doing it requires jumping through hoops (namely setting `_show_breadcrumbs = False` or using the template independently of the view classes) that should signal to any reasonable developer that what they're doing is unsupported, and C) the burden of fixing it by switching to the breadcrumb variant (or if that's really not possible, creating your own template based on whatever legacy version of wagtailadmin/generic/index.html you like) seems pretty small - I think it should be pretty uncontroversial to go ahead with this cleanup!

**Pull request merged**